website/src/components/Visualisations/Roles/Roles.tsx

```diff
@@ -1,14 +1,174 @@
-import React from "react";
-import logo from "../../../assets/logo/logo512.png";
+import React, { useState, useEffect } from "react";
 import styles from "./Roles.module.css";
+import {
+    BarChart,
+    Bar,
+    XAxis,
+    YAxis,
+    Tooltip,
+    Legend,
+    ResponsiveContainer,
+} from "recharts";
+import { TurnsInRoles, ProcessedRoleData } from "./Util/RoleTypes";
+import { getProcessedCIRoleData, processRoleData } from "./Util/ProcessedRoleData";
+import { loadLocalOutput } from "../../NewRun/utils";
+
+type CustomTooltipProps = {
+    active: boolean;
+    payload: [{ name: string; value: number; unit: string }];
+    label: string;
+    data: ProcessedRoleData;
+};
+
+const CustomTooltip = ({
+    active,
+    payload,
+    label,
+    data,
+}: CustomTooltipProps) => {
+    const getTurnsInRoles = (name: string): TurnsInRoles => {
+        const roles = data.find((d) => d.name === name)?.roles;
+        if (roles) {
+            const turnsAsPresident = roles.reduce(
+                (acc, a) => acc + a.president,
+                0
+            );
+            const turnsAsJudge = roles.reduce((acc, a) => acc + a.judge, 0);
+            const turnsAsSpeaker = roles.reduce((acc, a) => acc + a.speaker, 0);
+            const turnsAsNone = roles.reduce((acc, a) => acc + a.none, 0);
+            return new TurnsInRoles(
+                turnsAsPresident,
+                turnsAsJudge,
+                turnsAsSpeaker,
+                turnsAsNone
+            );
+        } else {
+            console.log(`[VIS ROLE] Could not find ${name} in data...`);
+            return new TurnsInRoles();
+        }
+    };
+
+    if (active) {
+        const turnsInRoles = getTurnsInRoles(label);
+        const turns =
+            turnsInRoles.president +
+            turnsInRoles.judge +
+            turnsInRoles.speaker +
+            turnsInRoles.none;
+
+        return (
+            <div className={styles.customTooltip}>
+                <p className={styles.label}>{label}</p>
+                <p className={styles.content}>
+                    Turns as President: {turnsInRoles.president} (
+                    {((turnsInRoles.president * 100) / turns).toFixed(1)}%)
+                </p>
```

Comment:
fine but you can use some sik tricks to map from `[president, judge, speaker, none].map(....// sik trick here)`